SUP-115-Create-Stream-Application-Audit (#408)
Macrometacorp/docs: docs/cep/tutorials/create-stream-app.md

<ul>
<li>Stateless query: Queries that only consider currently incoming events when generating an output. e.g., filters</li>
<li>Stateful query: Queries that consider both currently incoming events as well as past events when generating an output. e.g., windows, sequences, patterns, etc.</li>

> <li>Partitions: Collections of stream definitions and queries separated from each other within a Stream worker for the purpose of processing events in parallel and in isolation</li>

Reviewer: ```suggestion
<li>Partitions: Collections of stream definitions and queries separated from each other within a stream worker for the purpose of processing events in parallel and in isolation</li>
```

</ul></td>
</tr>
</tbody>
</table>

Macrometa provide in-build source, sink and store explained in the later section of this document.

## Creating a Stream Worker

To create a stream worker follow the steps below:

1. Open the GUI. Click the **Stream Workers** tab.
1. Click **New Stream Worker** to define a new stream worker.
1. Type a **Name** for the stream worker. For example, `SweetProductionAnalysis`.
1. Type a **Description**.

> 1. Add the following sample stream worker.

Reviewer: ```suggestion
1. Add the following sample stream worker:
```

	```sql
	CREATE SOURCE SweetProductionStream WITH (type = 'database', collection='SweetProductionData', collection.type='DOC', replication.type='GLOBAL',  map.type='json') (name string, amount double);
	
	CREATE SINK ProductionAlertStream WITH (type= 'stream', stream='ProductionAlertStream', map.type='json') (name string, amount double);
	
	INSERT INTO ProductionAlertStream
	SELECT *
	FROM SweetProductionStream;
	```

> 1. Click `Save` to save the stream worker.

Reviewer: ```suggestion
1. Click **Save** to save the stream worker.
```

> 1. Select all the regions to deploy your application in.

Reviewer: Is there a minimum of two here? I know there is a minimum for fabrics.

Author: A client can have just 2 in playground active, but many more stream workers not enabled

> 1. Click on `Save`.

Reviewer: ```suggestion
1. Click **Save**.
```

## Source

For this example, a source can be of type stream or database. The incoming data is used for processing. A source-of-type database is best if you need to store incoming data. Otherwise, you can use a source-type stream.

> ### Create Streams

Reviewer: ```suggestion

### Create Streams
```

Syntax:

> ```sql

Reviewer: ```suggestion

```sql
```

   CREATE SOURCE SourceName WITH (type="stream", stream.list="STRING", replication.type="STRING", map.type='type') (strings);
```

> Example:

Reviewer: ```suggestion

Example:
```

> ### C8Streams

Reviewer: ```suggestion
### Create Streams
```

</tbody>
</table>

### Create collections

You can create collections with your stream worker, and store incoming data in it for further processing, the syntax to achieve that is shown below:

> Syntax:

Reviewer: ```suggestion
Syntax:

```

```sql
   CREATE SOURCE SourceName WITH (type="database", collection="STRING", replication.type="STRING", collection.type="STRING", map.type='type') (strings);
```

> Example:

Reviewer: ```suggestion

Example:
```

> ### C8DB

Reviewer: It seems like this is the wrong heading. Maybe something about stream workers?

Syntax:
```sql
   CREATE SOURCE SourceName WITH (type="database", collection="STRING", replication.type="STRING", collection.type="STRING", map.type='type') (strings);
```

Example:
```sql
   CREATE SOURCE SweetProductionStream WITH (type = 'database', collection='SweetProductionData', collection.type='DOC', replication.type='GLOBAL',  map.type='json') (name string, amount double);
```

> Query Parameters:

Reviewer: ```suggestion
Query parameters:
```

> ### C8DB

Reviewer: Avoid stacked headings, need an intro paragraph after the Table heading above. Also, I don't think we're using C8DB anymore, so this seems like the wrong heading text for this section.

Syntax:
```sql
   CREATE STORE StoreName WITH (type="database", collection="STRING", replication.type="STRING", collection.type="STRING", map.type='type', from="STRING", to="STRING") (strings);
```
Example:
```sql
   CREATE STORE SweetProductionCollection WITH (type="database", collection="SweetProductionCollection", replication.type="local", collection.type="DOC", map.type='json') (name string, amount double);
```
Or equivalent using TABLE:
```sql
   CREATE TABLE GLOBAL SweetProductionCollection(name string, amount double);
```

> Stream applications will use the c8db with the default query parameters explained in the chart below.

Reviewer: Two things:
1. In all cases, replace "stream applications" with "stream workers." They used to call them stream applications, they don't anymore.
2. "C8DB" is also outdated terminology. Replace with "Macrometa" or the exact name of what is being accessed, like a collection.

## Table

Table is similar to collection, is a structured representation of data with a defined schema.

Syntax:

```sql
   CREATE TABLE GLOBAL TableName(property type);
```
Example:
```sql
   CREATE TABLE GLOBAL SweetProductionCollection(name string, amount double);
```

Or equivalent using STORE:
```sql
   CREATE STORE SweetProductionCollection WITH (type="database", collection="SweetProductionCollection", replication.type="local", collection.type="DOC", map.type='json') (name string, amount double);
```

> Stream worker will use the macrometa collections with the default query parameters explained in the chart below.

Reviewer: ```suggestion
The stream worker will use the Macrometa collections with the default query parameters explained in the chart below.
```

</tr>
<tr class="even">
<td>replication.type</td>

> <td>Specifies if the replication type of the c8db collection. At the moment local collections are not allowed, type must be `global`</td>

Reviewer: ```suggestion
<td>Specifies if the replication type of the collection. <b>Note:</b> Type must be `global`. Local collections are not currently allowed.</td>
```
